File: roisa/gui/image_list_widget.py
"""
image_list_widget.py — Compact panel listing all loaded images with
eye-toggle visibility, REF/IN badges, click-to-activate, and remove buttons.

Signals
-------
addRequested        ()              user clicked "+ Add"
activateRequested   (int)           user clicked a row name → make it active
visibilityToggled   (int, bool)     user clicked the eye toggle
removeRequested     (int)           user clicked ✕
"""

from __future__ import annotations

import os
from typing import Optional

from PyQt6.QtCore    import Qt, pyqtSignal
from PyQt6.QtWidgets import (
    QFrame, QHBoxLayout, QLabel, QPushButton, QSizePolicy,
    QVBoxLayout, QWidget,
)


# ── Style constants ───────────────────────────────────────────────────────────

_EYE_ON  = "●"
_EYE_OFF = "○"
_REMOVE  = "✕"

def _eye_style(on: bool) -> str:
    colour = "#6af" if on else "#445"
    return (f"QPushButton{{color:{colour};background:transparent;"
            f"border:none;font-size:11px;}}")

_REF_BADGE_STYLE = (
    "QLabel{background:#1a5fb4;color:#fff;border-radius:3px;"
    "padding:1px 3px;font-size:8px;font-weight:bold;}")

_IN_BADGE_STYLE = (
    "QLabel{background:#c06000;color:#fff;border-radius:3px;"
    "padding:1px 3px;font-size:8px;font-weight:bold;}")

def _name_style(active: bool) -> str:
    colour = "#eee" if active else "#aaa"
    hover  = "#fff" if active else "#ddd"
    return (f"QPushButton{{color:{colour};background:transparent;border:none;"
            f"font-size:10px;text-align:left;padding:0 2px;}}"
            f"QPushButton:hover{{color:{hover};}}")

_REMOVE_STYLE = (
    "QPushButton{color:#844;background:transparent;border:none;font-size:9px;}"
    "QPushButton:hover{color:#f66;}"
    "QPushButton:disabled{color:#2a2a2a;}")

def _row_style(active: bool, enabled: bool) -> str:
    if not enabled:
        return "QWidget{background:#111;border-radius:3px;}"
    if active:
        return "QWidget{background:#1a3d5c;border-radius:3px;}"
    return "QWidget{background:transparent;}"


# ── Row dataclass ─────────────────────────────────────────────────────────────

class _Row:
    __slots__ = ("widget", "eye_btn", "badge", "name_btn", "remove_btn", "enabled")

    def __init__(self, widget: QWidget, eye_btn: QPushButton,
                 badge: QLabel, name_btn: QPushButton,
                 remove_btn: QPushButton) -> None:
        self.widget     = widget
        self.eye_btn    = eye_btn
        self.badge      = badge
        self.name_btn   = name_btn
        self.remove_btn = remove_btn
        self.enabled    = True


# ═════════════════════════════════════════════════════════════════════════════

class ImageListWidget(QWidget):
    addRequested       = pyqtSignal()
    activateRequested  = pyqtSignal(int)
    visibilityToggled  = pyqtSignal(int, bool)
    removeRequested    = pyqtSignal(int)

    def __init__(self, parent: Optional[QWidget] = None) -> None:
        super().__init__(parent)
        self._rows:       list[_Row] = []
        self._active_idx: int        = -1

        outer = QVBoxLayout(self)
        outer.setContentsMargins(4, 4, 4, 2)
        outer.setSpacing(3)

        # ── Header: "Images" + "+ Add" ────────────────────────────────────────
        hdr  = QWidget(self)
        hdrL = QHBoxLayout(hdr)
        hdrL.setContentsMargins(0, 0, 0, 0)
        hdrL.setSpacing(4)

        title_lbl = QLabel("Images", hdr)
        title_lbl.setStyleSheet(
            "color:#9a9; font-size:10px; font-weight:bold;")
        hdrL.addWidget(title_lbl)
        hdrL.addStretch()

        add_btn = QPushButton("＋ Add", hdr)
        add_btn.setFixedHeight(18)
        add_btn.setStyleSheet(
            "QPushButton{background:#1a3d1a;color:#8c8;border:1px solid #3a6a3a;"
            "border-radius:2px;padding:1px 8px;font-size:10px;}"
            "QPushButton:hover{background:#246024;color:#afa;}")
        add_btn.clicked.connect(self.addRequested)
        hdrL.addWidget(add_btn)
        outer.addWidget(hdr)

        # ── Separator ─────────────────────────────────────────────────────────
        sep = QFrame(self)
        sep.setFrameShape(QFrame.Shape.HLine)
        sep.setStyleSheet("border:none;background:#2a2a2a;max-height:1px;")
        outer.addWidget(sep)

        # ── Row container ─────────────────────────────────────────────────────
        self._list_container = QWidget(self)
        self._list_container.setStyleSheet(
            "QWidget{background:#181818;border:1px solid #2a2a2a;"
            "border-radius:3px;}")
        self._list_layout = QVBoxLayout(self._list_container)
        self._list_layout.setContentsMargins(2, 2, 2, 2)
        self._list_layout.setSpacing(1)
        outer.addWidget(self._list_container)

        self._show_empty_placeholder()

    # ── Placeholder ───────────────────────────────────────────────────────────

    def _show_empty_placeholder(self) -> None:
        lbl = QLabel("No images loaded — use File → Open", self)
        lbl.setStyleSheet("color:#444; font-size:9px; padding:6px 4px;")
        lbl.setAlignment(Qt.AlignmentFlag.AlignCenter)
        lbl.setObjectName("_placeholder")
        self._list_layout.addWidget(lbl)

    def _remove_placeholder(self) -> None:
        for i in range(self._list_layout.count()):
            item = self._list_layout.itemAt(i)
            if item and item.widget() and \
                    item.widget().objectName() == "_placeholder":
                w = self._list_layout.takeAt(i).widget()
                w.deleteLater()
                return

    # ── Public API ────────────────────────────────────────────────────────────

    def clear(self) -> None:
        """Remove all rows and show the placeholder."""
        self._rows.clear()
        self._active_idx = -1
        while self._list_layout.count():
            item = self._list_layout.takeAt(0)
            if item.widget():
                item.widget().deleteLater()
            del item
        self._show_empty_placeholder()

    def add_image(self, display_name: str, is_ref: bool = False) -> None:
        """Append one row. is_ref=True → blue REF badge; else orange INn badge."""
        if len(self._rows) == 0:
            self._remove_placeholder()

        idx = len(self._rows)

        row = QWidget(self)
        row_layout = QHBoxLayout(row)
        row_layout.setContentsMargins(3, 1, 3, 1)
        row_layout.setSpacing(3)

        # Eye toggle
        eye_btn = QPushButton(_EYE_ON, row)
        eye_btn.setFixedSize(16, 16)
        eye_btn.setFlat(True)
        eye_btn.setStyleSheet(_eye_style(True))
        eye_btn.setToolTip("Toggle visibility")
        row_layout.addWidget(eye_btn)

        # Badge
        badge_text = "REF" if is_ref else f"IN{idx}"
        badge = QLabel(badge_text, row)
        badge.setFixedWidth(28)
        badge.setAlignment(Qt.AlignmentFlag.AlignCenter)
        badge.setStyleSheet(_REF_BADGE_STYLE if is_ref else _IN_BADGE_STYLE)
        row_layout.addWidget(badge)

        # Name button (flat, click → activate)
        short = os.path.basename(display_name) or display_name
        if len(short) > 22:
            short = short[:10] + "…" + short[-9:]
        name_btn = QPushButton(short, row)
        name_btn.setFlat(True)
        name_btn.setToolTip(display_name)
        name_btn.setSizePolicy(
            QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Preferred)
        name_btn.setStyleSheet(_name_style(False))
        row_layout.addWidget(name_btn, 1)

        # Remove button
        rm_btn = QPushButton(_REMOVE, row)
        rm_btn.setFixedSize(16, 16)
        rm_btn.setFlat(True)
        rm_btn.setStyleSheet(_REMOVE_STYLE)
        rm_btn.setToolTip(
            "Cannot remove reference while inputs exist" if is_ref
            else "Remove this input image")
        row_layout.addWidget(rm_btn)

        # Wire signals (capture idx by closure)
        def _eye_clicked(checked=False, _i=idx):
            self.visibilityToggled.emit(_i, not self._rows[_i].enabled)
        def _name_clicked(checked=False, _i=idx):
            self.activateRequested.emit(_i)
        def _rm_clicked(checked=False, _i=idx):
            self.removeRequested.emit(_i)

        eye_btn.clicked.connect(_eye_clicked)
        name_btn.clicked.connect(_name_clicked)
        rm_btn.clicked.connect(_rm_clicked)

        r = _Row(row, eye_btn, badge, name_btn, rm_btn)
        self._rows.append(r)
        self._list_layout.addWidget(row)

    def set_active(self, idx: int) -> None:
        """Highlight a row as the currently viewed image."""
        prev = self._active_idx
        self._active_idx = idx
        if 0 <= prev < len(self._rows):
            self._apply_row_style(prev)
        if 0 <= idx < len(self._rows):
            self._apply_row_style(idx)

    def set_enabled(self, idx: int, on: bool) -> None:
        """Update the eye glyph and dim/undim a row."""
        if idx < 0 or idx >= len(self._rows):
            return
        row = self._rows[idx]
        row.enabled = on
        row.eye_btn.setText(_EYE_ON if on else _EYE_OFF)
        row.eye_btn.setStyleSheet(_eye_style(on))
        self._apply_row_style(idx)

    def set_remove_enabled(self, idx: int, on: bool) -> None:
        """Enable or disable the ✕ button for a row."""
        if 0 <= idx < len(self._rows):
            self._rows[idx].remove_btn.setEnabled(on)

    # ── Private ───────────────────────────────────────────────────────────────

    def _apply_row_style(self, idx: int) -> None:
        if idx < 0 or idx >= len(self._rows):
            return
        row    = self._rows[idx]
        active = (idx == self._active_idx)
        row.widget.setStyleSheet(_row_style(active, row.enabled))
        row.name_btn.setStyleSheet(_name_style(active))

File: src/gui/ImageListWidget.cpp
// ImageListWidget.cpp

#include "ImageListWidget.h"

#include <QFileInfo>
#include <QFrame>
#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QVBoxLayout>

// ── Style constants ───────────────────────────────────────────────────────────

static const char* kEyeOn     = "●";
static const char* kEyeOff    = "○";
static const char* kRemove    = "✕";

static QString eyeBtnStyle(bool on) {
    return on
        ? "QPushButton{color:#6af;background:transparent;border:none;font-size:11px;}"
        : "QPushButton{color:#445;background:transparent;border:none;font-size:11px;}";
}
static QString refBadgeStyle() {
    return "QLabel{background:#1a5fb4;color:#fff;border-radius:3px;"
           "padding:1px 3px;font-size:8px;font-weight:bold;}";
}
static QString inBadgeStyle() {
    return "QLabel{background:#c06000;color:#fff;border-radius:3px;"
           "padding:1px 3px;font-size:8px;font-weight:bold;}";
}
static QString nameBtnStyle(bool active) {
    return active
        ? "QPushButton{color:#eee;background:transparent;border:none;"
          "font-size:10px;text-align:left;padding:0 2px;}"
          "QPushButton:hover{color:#fff;}"
        : "QPushButton{color:#aaa;background:transparent;border:none;"
          "font-size:10px;text-align:left;padding:0 2px;}"
          "QPushButton:hover{color:#ddd;}";
}
static QString removeBtnStyle() {
    return "QPushButton{color:#844;background:transparent;border:none;font-size:9px;}"
           "QPushButton:hover{color:#f66;}"
           "QPushButton:disabled{color:#2a2a2a;}";
}
static QString rowStyle(bool active) {
    return active
        ? "QWidget{background:#1a3d5c;border-radius:3px;}"
        : "QWidget{background:transparent;}";
}

// ═════════════════════════════════════════════════════════════════════════════

ImageListWidget::ImageListWidget(QWidget* parent) : QWidget(parent)
{
    auto* outer = new QVBoxLayout(this);
    outer->setContentsMargins(4, 4, 4, 2);
    outer->setSpacing(3);

    // ── Header: "Images" label + "+ Add" button ───────────────────────────────
    auto* hdr  = new QWidget(this);
    auto* hdrL = new QHBoxLayout(hdr);
    hdrL->setContentsMargins(0, 0, 0, 0);
    hdrL->setSpacing(4);

    auto* titleLbl = new QLabel("Images", hdr);
    titleLbl->setStyleSheet("color:#9a9; font-size:10px; font-weight:bold;");
    hdrL->addWidget(titleLbl);
    hdrL->addStretch();

    auto* addBtn = new QPushButton("＋ Add", hdr);
    addBtn->setFixedHeight(18);
    addBtn->setStyleSheet(
        "QPushButton{background:#1a3d1a;color:#8c8;border:1px solid #3a6a3a;"
        "border-radius:2px;padding:1px 8px;font-size:10px;}"
        "QPushButton:hover{background:#246024;color:#afa;}");
    connect(addBtn, &QPushButton::clicked, this, &ImageListWidget::addRequested);
    hdrL->addWidget(addBtn);
    outer->addWidget(hdr);

    // ── Thin separator ────────────────────────────────────────────────────────
    auto* sep = new QFrame(this);
    sep->setFrameShape(QFrame::HLine);
    sep->setStyleSheet("border:none;background:#2a2a2a;max-height:1px;");
    outer->addWidget(sep);

    // ── Scrollable row area ────────────────────────────────────────────────────
    auto* listContainer = new QWidget(this);
    listContainer->setStyleSheet(
        "QWidget{background:#181818;border:1px solid #2a2a2a;border-radius:3px;}");
    m_listLayout = new QVBoxLayout(listContainer);
    m_listLayout->setContentsMargins(2, 2, 2, 2);
    m_listLayout->setSpacing(1);
    outer->addWidget(listContainer);

    // Start with placeholder
    auto* empty = new QLabel("No images loaded — use File → Open", listContainer);
    empty->setStyleSheet("color:#444; font-size:9px; padding:6px 4px;");
    empty->setAlignment(Qt::AlignCenter);
    m_listLayout->addWidget(empty);
}

// ── Population ────────────────────────────────────────────────────────────────

void ImageListWidget::clear()
{
    m_rows.clear();
    m_activeIdx = -1;

    while (QLayoutItem* item = m_listLayout->takeAt(0)) {
        if (QWidget* w = item->widget()) w->deleteLater();
        delete item;
    }
}

void ImageListWidget::addImage(const QString& displayName, bool isRef)
{
    const int idx = m_rows.size();

    auto* row  = new QWidget(this);
    auto* rowL = new QHBoxLayout(row);
    rowL->setContentsMargins(3, 1, 3, 1);
    rowL->setSpacing(3);

    // Eye toggle
    auto* eyeBtn = new QPushButton(kEyeOn, row);
    eyeBtn->setFixedSize(16, 16);
    eyeBtn->setFlat(true);
    eyeBtn->setStyleSheet(eyeBtnStyle(true));
    eyeBtn->setToolTip("Toggle visibility");
    rowL->addWidget(eyeBtn);

    // Badge
    QString badgeText = isRef ? "REF" : QString("IN%1").arg(idx);
    auto* badge = new QLabel(badgeText, row);
    badge->setFixedWidth(28);
    badge->setAlignment(Qt::AlignCenter);
    badge->setStyleSheet(isRef ? refBadgeStyle() : inBadgeStyle());
    rowL->addWidget(badge);

    // Name (flat clickable button)
    QString shortName = QFileInfo(displayName).fileName();
    if (shortName.isEmpty()) shortName = displayName;
    if (shortName.length() > 22) shortName = shortName.left(10) + "…" + shortName.right(9);

    auto* nameBtn = new QPushButton(shortName, row);
    nameBtn->setFlat(true);
    nameBtn->setToolTip(displayName);
    nameBtn->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
    nameBtn->setStyleSheet(nameBtnStyle(false));
    rowL->addWidget(nameBtn, 1);

    // Remove button
    auto* rmBtn = new QPushButton(kRemove, row);
    rmBtn->setFixedSize(16, 16);
    rmBtn->setFlat(true);
    rmBtn->setStyleSheet(removeBtnStyle());
    rmBtn->setToolTip(isRef ? "Cannot remove reference while inputs exist"
                             : "Remove this input image");
    rowL->addWidget(rmBtn);

    // Connections (capture idx by value)
    connect(eyeBtn, &QPushButton::clicked, this, [this, idx]{
        const bool cur = m_rows[idx].enabled;
        emit visibilityToggled(idx, !cur);
    });
    connect(nameBtn, &QPushButton::clicked, this, [this, idx]{
        emit activateRequested(idx);
    });
    connect(rmBtn, &QPushButton::clicked, this, [this, idx]{
        emit removeRequested(idx);
    });

    Row r; r.widget=row; r.eyeBtn=eyeBtn; r.badge=badge;
           r.nameBtn=nameBtn; r.removeBtn=rmBtn; r.enabled=true;
    m_rows.append(r);
    m_listLayout->addWidget(row);
}

// ── State setters ─────────────────────────────────────────────────────────────

void ImageListWidget::setActive(int idx)
{
    const int prev = m_activeIdx;
    m_activeIdx    = idx;
    if (prev >= 0 && prev < m_rows.size()) applyRowStyle(prev);
    if (idx  >= 0 && idx  < m_rows.size()) applyRowStyle(idx);
}

void ImageListWidget::setEnabled(int idx, bool on)
{
    if (idx < 0 || idx >= m_rows.size()) return;
    m_rows[idx].enabled = on;
    m_rows[idx].eyeBtn->setText(on ? kEyeOn : kEyeOff);
    m_rows[idx].eyeBtn->setStyleSheet(eyeBtnStyle(on));
    // Dim the whole row slightly when hidden
    m_rows[idx].nameBtn->setStyleSheet(
        nameBtnStyle(idx == m_activeIdx));
    m_rows[idx].widget->setStyleSheet(
        on ? rowStyle(idx == m_activeIdx)
           : "QWidget{background:#111;border-radius:3px;}");
}

void ImageListWidget::setRemoveEnabled(int idx, bool on)
{
    if (idx >= 0 && idx < m_rows.size())
        m_rows[idx].removeBtn->setEnabled(on);
}

// ── Private ───────────────────────────────────────────────────────────────────

void ImageListWidget::applyRowStyle(int idx)
{
    if (idx < 0 || idx >= m_rows.size()) return;
    const bool active  = (idx == m_activeIdx);
    const bool enabled = m_rows[idx].enabled;
    m_rows[idx].widget->setStyleSheet(
        !enabled      ? "QWidget{background:#111;border-radius:3px;}"
        : active      ? rowStyle(true)
                      : rowStyle(false));
    m_rows[idx].nameBtn->setStyleSheet(nameBtnStyle(active));
}
